lms/migrations/versions/f3d631c110bf_canvas_files_enabled.py

```diff
@@ -0,0 +1,88 @@
+"""
+Make canvas files enabled an AI setting.
+
+Revision ID: f3d631c110bf
+Revises: 0b7391f5b1e3
+Create Date: 2022-12-28 10:15:40.830444
+
+"""
+import sqlalchemy as sa
+from alembic import op
+
+revision = "f3d631c110bf"
+down_revision = "5f0057d6d60f"
+
+
+def upgrade():
+    """
+    Explicitly store the `files_enabled` setting for canvas instances.
+
+    Currently the logic for this "setting" is:
+
+    ```
+    enabled = (request.product.family == Product.Family.CANVAS) and (
+        "custom_canvas_course_id" in request.lti_params
+        and application_instance.developer_key is not None
+    )
+    ```
+
+    Querying for just:
+
+        SELECT COUNT(id) FROM application_instances WHERE developer_key IS NOT NULL
+
+    returns 957 AIs at the time of writing
+
+    while trying to get the 'custom_canvas_course_id' part:
+
+        SELECT COUNT(id)
+        FROM application_instances
+        WHERE
+            developer_key IS NOT NULL
+            AND EXISTS (
+                SELECT id
+                FROM grouping
+                WHERE
+                    type = 'course'
+                    AND grouping.application_instance_id = application_instances.id
+                    AND extra->'canvas'->>'custom_canvas_course_id' IS NOT NULL
+            )
+
+    we get 491 AIs
+
+    but removing the last condition, the one with custom_canvas_course_id gives us the same number.
+
+    This means that all canvas instances that have a developer key and have created courses do send custom_canvas_course_id.
+    """
+    conn = op.get_bind()
+    result = conn.execute(
+        """
+        UPDATE application_instances
+            SET settings = CASE
+                WHEN settings = '{}' THEN '{"canvas": {"files_enabled": true}}'
+                WHEN settings->'canvas' is null  THEN jsonb_set(settings, '{canvas}', '{"files_enabled": true}')
+                ELSE jsonb_set(settings, '{canvas,files_enabled}', 'true')
+            END
+        WHERE
+            developer_key IS NOT NULL
+            AND (
+                -- Check we have records for custom_canvas_course_id
+                EXISTS (
+                    SELECT id
+                    FROM grouping
+                    WHERE
+                        type = 'course'
+                        AND grouping.application_instance_id = application_instances.id
+                        AND extra->'canvas'->>'custom_canvas_course_id' IS NOT NULL
+                )
+                -- Or has any other canvas feature enabled
+                OR settings->'canvas'->>'sections_enabled'= 'true'
+                OR settings->'canvas'->>'groups_enabled'= 'true'
```

Comment:
Using other features as a proxy

Reply:
It's tempting to turn this on for everyone, but I suppose that robs it of any value for analytic purposes.

This would be a great candidate to allow a customer to enable and disable it themselves.